File: packages/native/runtime-swift/Sources/PyreonRuntime/PyreonLinking.swift
// PyreonLinking — the SwiftUI side of Pyreon's external-URL service
// (M3.2b). Mirrors the core `@pyreon/hooks` `useLinking` shape and the
// Kotlin `PyreonLinking` one-for-one.
//
// Surface:
//
//     linking.openUrl("https://pyreon.dev")   // hand the URL to the OS
//
// `UIApplication.shared.open(_:)` asks the system to open the URL in the
// app registered for its scheme (Safari for http/https), backgrounding the
// current app. No presenting view controller and no permission needed —
// simpler than PyreonShare's UIActivityViewController.
//
// UIKit-only: the SPM package targets iOS 17 AND macOS 14, so the UIKit
// call is guarded (`#if canImport(UIKit)`); macOS/Linux builds compile to a
// no-op (matching how PyreonShare / PyreonClipboard branch). No third-party
// dependency.

import Foundation

#if canImport(UIKit)
import UIKit
#endif

/// External-URL opener — the SwiftUI half of `useLinking`.
public final class PyreonLinking {
    public init() {}

    /// Open `url` in the platform browser / the app registered for its
    /// scheme. An unparseable string is ignored.
    public func openUrl(_ url: String) {
        #if canImport(UIKit)
        guard let parsed = URL(string: url) else { return }
        UIApplication.shared.open(parsed, options: [:], completionHandler: nil)
        #endif
    }
}

File: packages/fundamentals/hooks/native/kotlin/com/pyreon/runtime/PyreonLinking.kt
// PyreonLinking — the Compose side of Pyreon's external-URL service
// (M3.2b). Mirrors the Swift `PyreonLinking` one-for-one + the core
// `@pyreon/hooks` `useLinking` shape.
//
// Surface:
//
//     linking.openUrl("https://pyreon.dev")   // hand the URL to the OS
//
// Android opens a URL with an `ACTION_VIEW` intent — the system routes it
// to the browser (or the app registered for the scheme). The `Context` is
// captured at CONSTRUCTION time (not per-call) so the public method matches
// Swift's one-for-one. PMTC emits
// `remember { PyreonLinking(LocalContext.current) }`.

package com.pyreon.runtime

import android.content.Context
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri

/**
 * External-URL opener — the Compose half of `useLinking`. Lives as
 * `remember { PyreonLinking(LocalContext.current) }` inside a composable.
 * No reactive state; `openUrl` fires an ACTION_VIEW intent.
 */
class PyreonLinking(private val context: Context) {
    /**
     * Open `url` in the platform browser / the app registered for its scheme.
     *
     * Returns whether anything could handle it. `startActivity` THROWS
     * `ActivityNotFoundException` when no app resolves the scheme, so an
     * unguarded call terminated the process on a URL the device simply could
     * not open — a `zoommtg://` link on a phone without Zoom, from a feed of
     * server-supplied links. The Swift half already degraded gracefully, so one
     * shared source crashed on exactly one target.
     *
     * Reported rather than swallowed: a caller that wants to fall back (open
     * the https equivalent, show a message) needs to know, and a bare `catch`
     * would leave the button looking broken with no way to tell.
     */
    fun openUrl(url: String): Boolean {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url)).apply {
            // startActivity may be called from a non-Activity Context; ACTION_VIEW
            // then needs NEW_TASK (harmless for an Activity Context too).
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        return try {
            context.startActivity(intent)
            true
        } catch (e: ActivityNotFoundException) {
            false
        }
    }
}
